tfsdk/type_plan_modification.go

```diff
@@ -0,0 +1,386 @@
+		err := state.As(&stateVals)
+		if err != nil {
+			diags.AddAttributeError(path, "Error generating plan", "An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nError converting state: "+err.Error())
+			return plan, diags
+		}
+	}
+	planVals := []tftypes.Value{}
+	err := plan.As(&planVals)
+	if err != nil {
+		diags.AddAttributeError(path, "Error generating plan", "An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nError converting plan: "+err.Error())
+		return plan, diags
+	}
+	elemTypes := wets.ElementTypes()
+	for index, elemPlan := range planVals {
+		if index >= len(elemTypes) {
+			diags.AddAttributeError(path, "Error generating plan", "An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nPlan has a value at "+path.WithElementKeyInt(index).String()+" that the tuple has no type for.")
+			return plan, diags
+		}
+		elemType := elemTypes[index]
+		if index >= len(stateVals) {
+			diags.AddAttributeError(path, "Error generating plan", "An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nPlan has a value at "+path.WithElementKeyInt(index).String()+" that has no state equivalent, which is not allowed in tuples.")
+			return plan, diags
+		}
+		elemState := tftypes.NewValue(elemType.TerraformType(ctx), nil)
+		if index < len(stateVals) {
+			elemState = stateVals[index]
+		}
+		newPlan, ds := attributeTypeModifyPlan(ctx, elemType, elemState, elemPlan, path.WithElementKeyInt(index))
+		diags.Append(ds...)
+		if diags.HasError() {
+			return plan, diags
+		}
+		rawNewPlan, err := newPlan.ToTerraformValue(ctx)
+		if err != nil {
+			diags.AddAttributeError(path, "Error generating plan",
+				fmt.Sprintf("An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nCouldn't convert the attr.Value at %s back to a tftypes.Value: %s", path.WithElementKeyInt(index), err),
+			)
+			return plan, diags
+		}
+		planVals[index] = rawNewPlan
+	}
+	err = tftypes.ValidateValue(typ.TerraformType(ctx), planVals)
+	if err != nil {
+		diags.AddAttributeError(path, "Error generating plan",
+			fmt.Sprintf("An unexpected error was encountered while trying to generate the plan. Please report the following to the provider developer:\n\nThe modified plan was no longer compatible with the Terraform type: %s", err),
+		)
+		return plan, diags
+	}
+	return tftypes.NewValue(typ.TerraformType(ctx), planVals), diags
+}
+
+func attributeTypeModifyPlan(ctx context.Context, typ attr.Type, state, plan tftypes.Value, path *tftypes.AttributePath) (attr.Value, diag.Diagnostics) {
+	var diags diag.Diagnostics
+	if plan.IsKnown() && !plan.IsNull() {
+		if typ.TerraformType(ctx).Is(tftypes.Object{}) {
```

Comment:
Nit: Seems this logic could be simplified via:

```go
tfType := typ.TerraformType(ctx)
switch {
case tfType.Is(tftypes.Object{}):
  plan, diags = attributeTypeModifyPlanObject(ctx, typ, state, plan, path)
case tfType.Is(tftypes.Map{}):
  plan, diags = attributeTypeModifyPlanMap(ctx, typ, state, plan, path)
// ...
}
if diags.HasError() {
  return plan, diags
}
```

Unless there's some explicit reason to return a `nil` value instead of the prior plan value during errors.